### PR TITLE
Refactor ProjectBuilding into Package Phases

### LINKED ISSUE #9499: De-duplicating buildInplaceUnpackedPackage and buildAndInstallUnpackedPackage
It seems that there are a lot of things in common between `buildInplaceUnpackedPackage` and `buildAndInstallUnpackedPackage`, however, there is also a lot of duplication between the two.

In my opinion, it would be better to factor the common bits (not changing any behaviour... best effort) and have a single function for both (similar) use cases. The subtle bugs due to the duplication and not doing things uniformly for the two situations seems much worse than possibly incurring a regression in the behaviour change, especially going forward.

### PR DESCRIPTION
**Pull request by alt-romes**
This refactor of Distribution.Client.ProjectBuilding does the following:

* Moves package file monitoring logic to Distribution.Client.ProjectBuilding.PackageFileMonitor

* Moves the `buildInplaceUnpackedPackage` and `buildAndInstallUnpackedPackage` with auxiliary functions to Distribution.Client.ProjectBuilding.UnpackedPackage

* Refactors the common bits of `buildInplaceUnpackedPackage` and `buildAndInstallUnpackedPackage` to remove considerable code duplication while simplifying and making both functions more structured.

Namely, to better structure build inplace vs build and install, I've introduced:

* `PackageBuildingPhase` describes the various phases of processing the unpacked package both inplace and on install * Configure * Build * Install (copy + register) * Test * Bench * Repl * Haddock

* Then, `buildAndRegisterUnpackedPackage` implements the common logic between the two functions (such as calls to ./Setup and the order of the phases) but delegates the logic specific to each phase to an argument function which maps `PackageBuildingPhase` to `IO` actions.

* Now, build inplace and build and install functions are comprised as: * A wrapper around `buildAndRegisterUnpackedPackage` which does things specific to each before and after the main phases are processed * A delegate function which maps an action to each package processing phase

Fixes #9499

### CODE REVIEW

**Comment by alt-romes** [comment]
Here we are @andreabedini

**Comment by andreabedini** [comment]
@alt-romes I'll look at this ASAP <3

**Comment by alt-romes** [comment]
@andreabedini I'm quite happy with how it turned out. FWIW I suggest reviewing locally with --color-moved because not that much changed.

OH, wait! I've been debugging locally 2 failures for the past hour... but CI is green?... Well, great!

**Comment by alt-romes** [comment]
@andreabedini the commit message might also help in guiding the review.

**Comment by Mikolaj** [comment]
@alt-romes: is this ready for review? If so, could you stick the needs-review label on it?

**Review by Mikolaj — APPROVED** [review]
@alt-romes: thank you. I think `--color-moved` doesn't quite work for me now due to the conflicts. But from a quick glimpse, the PR looks good.

**Comment by alt-romes** [comment]
I've rebased on master and solved the conflict @Mikolaj, it should be good to read now.
@andreabedini could you also take a look?

File: cabal-install/src/Distribution/Client/ProjectBuilding.hs.orig [hunk 1]
```diff
@@ -0,0 +1,2345 @@
+{-# LANGUAGE BangPatterns #-}
```
Comment (ulysses4ever) [inline cabal-install/src/Distribution/Client/ProjectBuilding.hs.orig:1]: Is this .orig file committed by mistake?

**Review by ulysses4ever — APPROVED** [review]
Thanks!

The only issue is the extra file, I think. (See below)

File: cabal-install/src/Distribution/Client/ProjectBuilding.hs.orig (hunk 1 above)
Comment (alt-romes) [reply cabal-install/src/Distribution/Client/ProjectBuilding.hs.orig]: Yes, it is. git creates these when resolving conflicts with `git mergetool -t vimdiff3`...
I've removed and pushed again. I'm assigning to merge.

**Comment by andreabedini** [comment]
@alt-romes Apologies for the radio silence. Holidays arrived and I missed the boat. This is fanstastic work :star:

**Comment by alt-romes** [comment]
Thanks @andreabedini

**Pull request merged**